### PR TITLE
[receiver/k8sobjects] ensure the `k8s.namespace.name` attribute is set for objects retrieved using the `watch` mode

### LINKED ISSUE #36352: [receiver/k8sobjects] Kubernetes events are missing namespace resource attribute
### Component(s)

receiver/k8sobjects

## Description

Kubernetes events, collected using the `watch` mode:
```
k8sobjects:
          objects:
            - name: events
              mode: "watch"
              group: "events.k8s.io"
              exclude_watch_type:
                - "DELETED"
```
are missing the namespace `resource.attribute`.
Looking to the code - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.112.0/receiver/k8sobjectsreceiver/unstructured_to_logdata.go#L58-L60 it seems that it is expected that this attribute is added.

`watchObjects` have a different [structure](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.112.0/receiver/k8sobjectsreceiver/unstructured_to_logdata.go#L25-L32) comparing to the `pullObjects` as a result `e.GetNamespace()` [call](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.112.0/receiver/k8sobjectsreceiver/unstructured_to_logdata.go#L58) return an empty string.

another option would be to add `namespace` to the `attributes` https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.112.0/receiver/k8sobjectsreceiver/unstructured_to_logdata.go#L36-L40, would it be a preferable way?

## Steps to Reproduce
enable k8sobjects:
```
k8sobjects:
          objects:
            - name: events
              mode: "watch"
              group: "events.k8s.io"
              exclude_watch_type:
                - "DELETED"
```

## Expected Result
I expect to see Resource Attribute `k8s.namespace.name` added

## Actual Result
```
ObservedTimestamp: 2024-11-13 15:27:52.068912453 +0000 UTC
Timestamp: 1970-01-01 00:00:00 +0000 UTC
SeverityText:
SeverityNumber: Unspecified(0)
Body: Map(...)
Attributes:
     -> k8s.resource.name: Str(events)
     -> event.domain: Str(k8s)
     -> event.name: Str(coredns-7db6d8ff4d-9hzj5.180790b119d77167)
```

### Collector version

0.112.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bacherfl**
#### Description
This PR ensures that the structure of log records generated by the k8sobjects receiver is the same, regardless of the mode (`watch` or `pull`) being used. This also solves the issue of the `k8s.namespace.name` attribute not being set for objects retrieved with `watch` mode. 

#### Link to tracking issue
Fixes #36352 

#### Testing

Added unit tests and adapted e2e tests

### CODE REVIEW

File: receiver/k8sobjectsreceiver/unstructured_to_logdata.go [hunk 1]
```diff
@@ -37,6 +34,7 @@ func watchObjectsToLogData(event *watch.Event, observedAt time.Time, config *K8s
 		if name != "" {
 			attrs.PutStr("event.domain", "k8s")
 			attrs.PutStr("event.name", name)
+			attrs.PutStr("event.type", string(event.Type))
```
Comment (tetianakravchenko) [inline receiver/k8sobjectsreceiver/unstructured_to_logdata.go:37]: FYI: there is an issue regarding reflecting of k8s events to the events API definition - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35857#issuecomment-2420214811, maybe it makes sense to include `event.type` to the events api convention

File: .chloggen/k8sobjects-data-structure.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: k8sobjectsreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: use same data structure for objects received via `pull` and `watch` mode
```
Comment (ChrsMark) [inline .chloggen/k8sobjects-data-structure.yaml:10]: It is preferred change-log entries to be in the form of user facing messages about what is fixed, added etc without surfacing the technical details.

How about sth the following:

```suggestion
note: Ensure `pull` and `watch` modes provide same set of attributes and resource attributes.
```

File: receiver/k8sobjectsreceiver/testdata/e2e/collector/configmap.yaml [hunk 3]
```diff
@@ -29,10 +29,11 @@ data:
         log_statements:
           - context: log
             statements:
-              - replace_match(body["object"]["metadata"]["uid"], "*", "12345")
-              - replace_match(body["object"]["metadata"]["resourceVersion"], "*", "12345")
-              - replace_match(body["object"]["metadata"]["creationTimestamp"], "*", "2000-01-01T00:00:00Z")
-              - replace_match(body["object"]["metadata"]["managedFields"][0]["time"], "*", "2000-01-01T00:00:00Z")
+              - replace_match(body["metadata"]["uid"], "*", "12345")
```
Comment (ChrsMark) [inline receiver/k8sobjectsreceiver/testdata/e2e/collector/configmap.yaml:32]: This looks like a breaking change to me. Is this something that we want to introduce?

**Review by ChrsMark — COMMENTED** [review]
Thank's @bacherfl for working on this!

I agree that both `pull` and `watch` modes should populate the `k8s.namespace.name` resource attribute as it was highlighted at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36352.

However, I'm not sure if the existing structure should change, `object.metadata.*` -> `metadata.*`.

Isn't this a breaking change? Should we ensure if we really want this or not? 
Otherwise, if we can fix the namespace issue without introducing a generic breaking change that would be optimal I think.

Could we clarify the above questions?

**Comment by bacherfl** [comment]
> Thank's @bacherfl for working on this!
> 
> I agree that both `pull` and `watch` modes should populate the `k8s.namespace.name` resource attribute as it was highlighted at #36352.
> 
> However, I'm not sure if the existing structure should change, `object.metadata.*` -> `metadata.*`.
> 
> Isn't this a breaking change? Should we ensure if we really want this or not? Otherwise, if we can fix the namespace issue without introducing a generic breaking change that would be optimal I think.
> 
> Could we clarify the above questions?

Thanks @ChrsMark for the feedback! You're right, changing the structure reported by the `watch` mode to be the same as with `pull` would be a breaking change, so I can also adapt the PR to specifically fix the namespace being set, if that is more preferable. In this case I would also add a note in the readme that the structure of the emitted log entries might differ, based on which mode is being used for the receiver

**Comment by bacherfl** [comment]
> > Thanks @ChrsMark for the feedback! You're right, changing the structure reported by the watch mode to be the same as with pull would be a breaking change, so I can also adapt the PR to specifically fix the namespace being set, if that is more preferable. In this case I would also add a note in the readme that the structure of the emitted log entries might differ, based on which mode is being used for the receiver
> 
> Cool! I think it's just fine fixing the namespace issue here and possibly file another issue to evaluate the "inconsistency" between the 2 modes. (in case we decide to change this, it should probably be done with a feature gate etc).
> 
> To my mind there should be no difference in the 2 modes from users' perspective but maybe I miss the details here behind this decision.

alright, then I'll adapt the current PR accordingly and will file an issue regarding the two different data structures - will let you know when this one is ready for another review :)

File: receiver/k8sobjectsreceiver/unstructured_to_logdata.go [hunk 4]
```diff
@@ -9,7 +9,7 @@ import (
 
 	"go.opentelemetry.io/collector/pdata/pcommon"
 	"go.opentelemetry.io/collector/pdata/plog"
-	semconv "go.opentelemetry.io/collector/semconv/v1.9.0"
+	semconv "go.opentelemetry.io/collector/semconv/v1.13.0"
```
Comment (ChrsMark) [inline receiver/k8sobjectsreceiver/unstructured_to_logdata.go:12]: Is this upgraded needed in this PR? If not, I suggest we do it in a standalone PR.

File: receiver/k8sobjectsreceiver/unstructured_to_logdata.go [hunk 5]
```diff
@@ -79,3 +79,15 @@ func unstructuredListToLogData(event *unstructured.UnstructuredList, observedAt
...
 	return out
 }
+
+func getNamespace(e unstructured.Unstructured) string {
+	// first, try to use the GetNamespace() method, which checks for the metadata.namespace property
+	if namespace := e.GetNamespace(); namespace != "" {
+		return namespace
+	}
+	// try to look up namespace in object.metadata.namespace (for objects reported via watch mode)
+	if namespace, ok, _ := unstructured.NestedString(e.Object, "object", "metadata", "namespace"); ok && namespace != "" {
```
[the remark this answers is not in the document]
Comment (bacherfl) [reply receiver/k8sobjectsreceiver/unstructured_to_logdata.go:89]: ah yes, good catch - that check is indeed not needed here, will remove it

File: receiver/k8sobjectsreceiver/unstructured_to_logdata.go (hunk 4 above)
Comment (bacherfl) [reply receiver/k8sobjectsreceiver/unstructured_to_logdata.go]: just checked and seems like this was not necessary - I reverted that change now

**Review by ChrsMark — APPROVED** [review]
lgtm

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**